[HUDI-687] Stop incremental reader on RO table before a pending compaction

satishkotha opened the pull request:
## What is the purpose of the pull request
example timeline:

t0 -> create bucket1.parquet
t1 -> create and append updates bucket1.log
t2 -> request compaction
t3 -> create bucket2.parquet

if compaction at t2 takes a long time, incremental reads using HoodieParquetInputFormat may make progress to read commits at t3 and skip data ingested at t1 leading to 'data loss' .(Data will still be on disk, but incremental readers wont see it because its in log file and readers move to t3)

To workaround this problem, we want to stop returning data belonging to commits > compaction_requested/inprogress_instant. After compaction is complete, incremental reader would see updates in t2, t3, so on. Disadvantage is that long running compactions can make it look like reader is 'stuck'. But that is better than skipping updates.

## Brief change log

- Change HoodieParquetInputFormat to read commits prior to compaction instant
- Added unit tests to validate behavior
- Fix broken test utils for reading records

## Verify this pull request
This change added tests and can be verified as follows:
mvn test (TestMergeOnReadTable and TestHoodieActiveTimeline)

Some discussion is on https://github.com/apache/incubator-hudi/pull/1389, sorry I messed up rebase, so resending as a new PR to avoid confusion

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

satishkotha commented:
@bvaradar sorry, I messed up rebase on https://github.com/apache/incubator-hudi/pull/1389, Please take a look at this instead. As discussed in the other PR, I updated RO and RT views. Spark DataSource does not seem to support MOR tables, so i'm skipping that part for now.

vinothchandar commented:
> if compaction at t2 takes a long time, incremental reads using HoodieParquetInputFormat may make progress to read commits at t3

IIUC this is because you are incremental pulling from the parquet only table? I thought we can already incremental pull via logs. no? cc @n3nash .. is this really needed since it will add complexity to the system..

Eventually, I would like incremental query/pull on MOR to be just based on logs..

satishkotha commented:
> > if compaction at t2 takes a long time, incremental reads using HoodieParquetInputFormat may make progress to read commits at t3
> 
> IIUC this is because you are incremental pulling from the parquet only table? I thought we can already incremental pull via logs. no? cc @n3nash .. is this really needed since it will add complexity to the system..
> 
> Eventually, I would like incremental query/pull on MOR to be just based on logs..

Based on view type, hudi decides the input format to use (see https://github.com/apache/incubator-hudi/blob/master/hudi-hive-sync/src/main/java/org/apache/hudi/hive/HiveSyncTool.java#L91) . For RO views, we use HoodieParquetInputFormat, which does not read log files. For RT views, we use HoodieParquetRealtimeInputFormat, which reads slice including log files. In my limited testing, incremental reads on RT views also do not work well (we see duplicates after compaction in some conditions).  @bvaradar  is working on fixing any broken windows for supporting incremental reads on RT views.

We wanted to include this change for supporting RO views (which is majority of use cases for us). I agree with you that this is additional complexity. I added more tests than usual because of that. 

Other alternatives i can think of:
1) Support incremental reads only for RT views.  incremental reads on RO can fail or use RT (is this your proposal in the above comment?)
2) Instead of doing incremental reads based on hoodie commit time, use parquet file creation times. This approach requires substantial changes and likely be breaking some fundamental assumptions.

Also, at a high level, I want to discuss adding additional mode for incremental reads. Today, its responsibility of hoodie users to save commit times and use that for next incremental reads. Can we add 'kafka consumer' model, where consumer only specifies their unique-id. Hudi tracks read progress (perhaps as part of consolidated metadata?). This would simplify usage and make debugging lot easier.
 
fyi,  @n3nash is out of office for next 10 days. @bvaradar likely can share more context. Let me know if you have other suggestions.

vinothchandar commented:
>>(is this your proposal in the above comment?)

yes.. Breaking away from _hoodie_commit_time is not a good approach, due to the reasons you cited. I agree there..

satishkotha commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTimeline.java:
```diff
@@ -144,6 +162,11 @@
+  /**
+   * Create a new Timeline with all instants before specified time.
+   */
+  HoodieTimeline findInstantsBefore(String time);
```
done

satishkotha commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTimeline.java:
```diff
@@ -127,6 +127,24 @@
+   *
+   * @return
+   */
+  HoodieTimeline filterInstantsBeforePendingCompaction();
```
Moved it to HoodieParquetInputFormat as suggested

satishkotha commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTimeline.java:
```diff
@@ -127,6 +127,24 @@
+  /**
+   * Get all instants (commits, delta commits) that produce new data, in the active timeline.
+   */
+  HoodieTimeline getCommitsTimeline();
```
I removed it

satishkotha commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/HoodieParquetInputFormat.java:
```diff
@@ -116,6 +117,24 @@
+   * To workaround this problem, we want to stop returning data belonging to commits > t2.
+   * After compaction is complete, incremental reader would see updates in t2, t3, so on.
+   */
+  protected HoodieTimeline filterInstantsTimeline(HoodieTimeline timeline) {
```
done

bvaradar commented on hudi-client/src/test/java/org/apache/hudi/common/HoodieMergeOnReadTestUtils.java:
```diff
@@ -47,27 +48,35 @@
-            newRecord.set(field, values[2]);
-          });
-          records.add(newRecord.build());
+        for (InputSplit split : splits) {
```
Was this working because these tests only dealt with 1 file-groups ? I can see count based assertions using records collected from this method

satishkotha replied:
yes, correct.

bvaradar commented on hudi-common/src/test/java/org/apache/hudi/common/table/string/TestHoodieActiveTimeline.java:
```diff
@@ -160,6 +159,8 @@ public void testTimelineOperations() {
         .filterCompletedInstants().findInstantsInRange("04", "11").getInstants().map(HoodieInstant::getTimestamp));
     HoodieTestUtils.assertStreamEquals("", Stream.of("09", "11"), timeline.getCommitTimeline().filterCompletedInstants()
         .findInstantsAfter("07", 2).getInstants().map(HoodieInstant::getTimestamp));
+    HoodieTestUtils.assertStreamEquals("", Stream.of("01", "03", "05"), timeline.getCommitTimeline().filterCompletedInstants()
```
nit: Add non-empty message for the assertion.

satishkotha replied:
fixed

bvaradar commented on hudi-client/src/test/java/org/apache/hudi/table/TestMergeOnReadTable.java:
```diff
@@ -80,6 +85,12 @@
 
 public class TestMergeOnReadTable extends HoodieClientTestHarness {
 
+  private HoodieParquetInputFormat roInputFormat;
```
We can add similar test-cases for Copy-On-Write table with RO inputformat alone to make it uniform. @satishkotha : Can you look at it as part of this PR or a follow-up ?

satishkotha replied:
Thanks for suggestion. I modified one of the existing tests on COW table. Please take a look.

satishkotha commented:
> @satishkotha : Some minor comments. Will approve once you reply/address them. Let's also wait for @vinothchandar to take a pass

@bvaradar Sounds good. I addressed your review comments.

satishkotha commented:
@vinothchandar could you take another look at this one?

vinothchandar commented:
if @bvaradar approves, it should be sufficient.. I will take another pass at it today

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieDefaultTimeline.java:
```diff
@@ -136,6 +136,13 @@ public HoodieDefaultTimeline findInstantsAfter(String instantTime, int numCommit
   }
 
+  @Override
+  public HoodieDefaultTimeline findInstantsBefore(String instantTime) {
```
Didn't realize you can have an overridden method with a return type which is a sub-type of original return type. Learnt something new today :)

satishkotha commented:
> Looks great @satishkotha Once you rebase and fix conflicts, we should be good to commit.

@vinothchandar Checking to see if you have any other feedback here?

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/HoodieParquetInputFormat.java:
```diff
@@ -118,6 +119,34 @@
+   */
+  protected HoodieDefaultTimeline filterInstantsTimeline(HoodieDefaultTimeline timeline) {
+    Option<HoodieInstant> pendingCompactionInstant = timeline.filterPendingCompactionTimeline().firstInstant();
+    if (pendingCompactionInstant.isPresent()) {
```
This seems like the crux of the change? and most of the other code is improving tests etc. If so, this seems like a  reasonable interim solution to me... Although we should encourage users to do incremental pull out of the RTInputFormat really ... 

The core problem of "data loss" being brought in this issue, feels like a mis-expectation really :)

satishkotha replied:
Yes, this is the crux of the change. My understanding is this bug caused data loss on derived ETL tables multiple times. These ETL tables are generated using incremental reads on "RO"views. As you suggested, that is core issue and switching to RT views is likely going to get rid of the problem. 

Also given "getting started" and other demo examples  include incremental reads on RO views,  I think this is useful to have especially given that finding root cause  for this took a while.

satishkotha replied:
I can introduce jobConf variable. I'm also agreeing with you that RT is the right approach. I'm just suggesting that we remove incremental read examples in different documents. For example, docker demo shows[ incremental reads on RO views](https://hudi.apache.org/docs/docker_demo.html#step-9-run-hive-queries-including-incremental-queries). So, people are likely to use this and end up with this difficult to debug problem.

Also, you are right about last statement. I already have tests to show that compaction timestamp is used for all updated records and not the update timestamp. Please see line 256 (last line in TestMergeOnReadTable#testIncrementalReadsWithCompaction)  that does not include updateTime. We validate that records include compactionCommitTime instead

vinothchandar replied:
Good point.. I think we can fix it forward since MOR incr pull issue is being fixed currently and next release, we can move the demo docs to work off _rt table..

vinothchandar commented on hudi-client/src/test/java/org/apache/hudi/table/TestMergeOnReadTable.java:
```diff
@@ -186,6 +168,96 @@ public void testSimpleInsertAndUpdate() throws Exception {
+  // test incremental read does not go past compaction instant for RO views
+  // For RT views, incremental read can go past compaction
+  @Test
+  public void testIncrementalReadsWithCompaction() throws Exception {
```
Let's make it clear that this an incremental on RO.. rename to `testIncrementalROReadsWithCOmpaction()`?

satishkotha replied:
This actually includes RT views too.

Example line 195: getRTIncrementalFiles(partitionPath); validateIncrementalFiles()

vinothchandar commented on hudi-client/src/test/java/org/apache/hudi/table/TestMergeOnReadTable.java:
```diff
@@ -1311,4 +1383,111 @@ private void assertNoWriteErrors(List<WriteStatus> statuses) {
     }
   }
+  
+  private FileStatus[] insertAndGetFilePaths(List<HoodieRecord> records, HoodieWriteClient client,
```
I am surprised by the amount of test helper code you needed to write yourself.. Should we be using code from some existing helpers.. or should these helpers move somewhere else, so they are beneficial over all?

satishkotha replied:
This one is refactored from existing test 'TestMergeOnReadTable#testSimpleInsertAndUpdate' . I don't know this setup is needed in other places. If you have any suggestions for new location, let me know.

There are also no tests so far with end-to-end testing of incremental reads on MOR tables (as far as i could tell). So I had to add some more utility test methods for that. Please suggest if ythere is a better location for that

vinothchandar replied:
let's have it like this for now.. Thanks for the explanation..!

vinothchandar commented:
@bhasudha is fixing the RT incremental pull. Hopefully this is all much easier..

vinothchandar commented:
@bvaradar Please take a final pass and merge.. I don't have concerns to move forward.. cc @satishkotha
